File: src/core/net/web/html-scan.test.ts
import { describe, expect, it } from 'vitest'

import { elementIdOf, findCalls, innerHtmlOf, parseObjectLiterals } from './html-scan'

/**
 * 扫描器的边界用例。整页反解的回归线在 `read-html.test.ts`（真实抓包），
 * 这里钉的是那份注释里说「所以不能用正则」的那几种情形——**都出自用户内容**，
 * 一旦踩中，表现是整页反解莫名其妙地少几楼或串行。
 */

describe('findCalls', () => {
  it('实参里的括号与引号不算数,按顶层逗号切', () => {
    const calls = findCalls("x.f( 1,$('a'),')',\"(\",null )", 'x.f(')
    expect(calls).toHaveLength(1)
    expect(calls[0]!.args).toEqual([
      { kind: 'number', value: 1 },
      { kind: 'expression', text: "$('a')" },
      { kind: 'string', value: ')' },
      { kind: 'string', value: '(' },
      { kind: 'null' },
    ])
  })

  it('转义引号不会提前收尾', () => {
    const calls = findCalls("x.f('it\\'s )', 2)", 'x.f(')
    expect(calls[0]!.args[0]).toEqual({ kind: 'string', value: "it's )" })
    expect(calls[0]!.args[1]).toEqual({ kind: 'number', value: 2 })
  })

  it('多处调用按出现顺序全找出来', () => {
    expect(findCalls('a.f(1) b.g(2) a.f(3)', 'a.f(').map((call) => call.args[0])).toEqual([
      { kind: 'number', value: 1 },
      { kind: 'number', value: 3 },
    ])
  })
})

describe('elementIdOf', () => {
  it("认 $('id') 这一种,别的实参一律 undefined", () => {
    expect(elementIdOf({ kind: 'expression', text: "$('postcontent7')" })).toBe('postcontent7')
    expect(elementIdOf({ kind: 'null' })).toBeUndefined()
    expect(elementIdOf({ kind: 'string', value: 'postcontent7' })).toBeUndefined()
  })
})

describe('innerHtmlOf', () => {
  it('同名标签嵌套时找到对的那个收尾', () => {
    const html = "<div><p id='c'>前<span>里</span>后</p></div>"
    expect(innerHtmlOf(html, 'c')).toBe('前<span>里</span>后')
    expect(innerHtmlOf("<span id='c'>a<span>b</span>c</span>d", 'c')).toBe('a<span>b</span>c')
  })

  it('自闭合标签不加深度(正文里的换行全是 `<br/>`)', () => {
    expect(innerHtmlOf("<p id='c'>一<br/>二</p>", 'c')).toBe('一<br/>二')
  })

  it('实体不解码:与 JSON 接口的 content 同口径', () => {
    expect(innerHtmlOf("<p id='c'>a&amp;b&lt;c</p>", 'c')).toBe('a&amp;b&lt;c')
  })

  it('id 前缀相同的元素不会张冠李戴', () => {
    const html = "<p id='postcontent1'>一</p><p id='postcontent12'>十二</p>"
    expect(innerHtmlOf(html, 'postcontent1')).toBe('一')
    expect(innerHtmlOf(html, 'postcontent12')).toBe('十二')
  })

  it('空元素与找不到的 id', () => {
    expect(innerHtmlOf("<h3 id='c'></h3>", 'c')).toBe('')
    expect(innerHtmlOf("<h3 id='c'></h3>", 'nope')).toBeUndefined()
  })
})

describe('parseObjectLiterals', () => {
  it('键不带引号的 JS 对象数组(附件表就是这个形态)', () => {
    expect(parseObjectLiterals("[{aid:'',url:'a.jpg',thumb:'56'},{url:'b.jpg',size:101}]")).toEqual([
      { aid: '', url: 'a.jpg', thumb: '56' },
      { url: 'b.jpg', size: '101' },
    ])
  })
})

File: app/src/test/kotlin/com/chasel/ng2n/core/net/web/HtmlScanTest.kt
package com.chasel.ng2n.core.net.web

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class HtmlScanTest {

  @Test
  fun `实参里的括号与引号不算数,按顶层逗号切`() {
    val calls = findCalls("x.f( 1,\$('a'),')',\"(\",null )", "x.f(")

    assertEquals(1, calls.size)
    assertEquals(
      listOf(
        JsArgument.Num(1.0),
        JsArgument.Expression("\$('a')"),
        JsArgument.Str(")"),
        JsArgument.Str("("),
        JsArgument.Null,
      ),
      calls[0].args,
    )
  }

  @Test
  fun `转义引号不会提前收尾`() {
    val calls = findCalls("x.f('it\\'s )', 2)", "x.f(")

    assertEquals(JsArgument.Str("it's )"), calls[0].args[0])
    assertEquals(JsArgument.Num(2.0), calls[0].args[1])
  }

  @Test
  fun `多处调用按出现顺序全找出来`() {
    val first = findCalls("a.f(1) b.g(2) a.f(3)", "a.f(").map { it.args[0] }

    assertEquals(listOf(JsArgument.Num(1.0), JsArgument.Num(3.0)), first)
  }

  @Test
  fun `只认 dollar 括号引号 id 这一种,别的实参一律 null`() {
    assertEquals("postcontent7", elementIdOf(JsArgument.Expression("\$('postcontent7')")))
    assertNull(elementIdOf(JsArgument.Null))
    assertNull(elementIdOf(JsArgument.Str("postcontent7")))
  }

  @Test
  fun `同名标签嵌套时找到对的那个收尾`() {
    assertEquals("前<span>里</span>后", innerHtmlOf("<div><p id='c'>前<span>里</span>后</p></div>", "c"))
    assertEquals("a<span>b</span>c", innerHtmlOf("<span id='c'>a<span>b</span>c</span>d", "c"))
  }

  @Test
  fun `自闭合标签不加深度(正文里的换行全是 br)`() {
    assertEquals("一<br/>二", innerHtmlOf("<p id='c'>一<br/>二</p>", "c"))
  }

  @Test
  fun `实体不解码,与 JSON 接口的 content 同口径`() {
    assertEquals("a&amp;b&lt;c", innerHtmlOf("<p id='c'>a&amp;b&lt;c</p>", "c"))
  }

  @Test
  fun `id 前缀相同的元素不会张冠李戴`() {
    val html = "<p id='postcontent1'>一</p><p id='postcontent12'>十二</p>"

    assertEquals("一", innerHtmlOf(html, "postcontent1"))
    assertEquals("十二", innerHtmlOf(html, "postcontent12"))
  }

  @Test
  fun `空元素与找不到的 id`() {
    assertEquals("", innerHtmlOf("<h3 id='c'></h3>", "c"))
    assertNull(innerHtmlOf("<h3 id='c'></h3>", "nope"))
  }

  @Test
  fun `键不带引号的 JS 对象数组(附件表就是这个形态)`() {
    val parsed = parseObjectLiterals("[{aid:'',url:'a.jpg',thumb:'56'},{url:'b.jpg',size:101}]")

    assertEquals(
      listOf(
        mapOf("aid" to "", "url" to "a.jpg", "thumb" to "56"),
        mapOf("url" to "b.jpg", "size" to "101"),
      ),
      parsed,
    )
  }

  @Test
  fun `匹配到空串与整支没匹配是两回事`() {
    assertEquals(listOf(mapOf("aid" to "", "n" to "-3")), parseObjectLiterals("{aid:'',n:-3}"))
  }

  @Test
  fun `balancedSlice 跳过字符串里的括号`() {
    val slice = balancedSlice("f({a:'(',b:')'})", 0, '{', '}')

    assertEquals("a:'(',b:')'", slice?.body)
  }

  @Test
  fun `readIntVariable 认 parseInt 包裹与裸赋值`() {
    assertEquals(45150945L, readIntVariable("__CURRENT_TID=45150945,", "__CURRENT_TID"))
    assertEquals(10000001L, readIntVariable("var __CURRENT_UID = parseInt('10000001',10),", "__CURRENT_UID"))
    assertNull(readIntVariable("nothing here", "__CURRENT_TID"))
  }

  @Test
  fun `readStringVariable 认转义引号`() {
    assertEquals("img.nga.cn", readStringVariable("__ATTACH_BASE_VIEW = 'img.nga.cn';", "__ATTACH_BASE_VIEW"))
    assertEquals("it's", readStringVariable("""x = 'it\'s';""", "x"))
  }

  @Test
  fun `readMarkedSection 取注释标记之间那段`() {
    assertEquals("2048", readMarkedSection("<!--msgcodestart-->2048<!--msgcodeend-->", "msgcode"))
    assertNull(readMarkedSection("<!--msgcodestart-->2048", "msgcode"))
  }
}
